### PR TITLE
ComputeImplementationForInterfaceMember shouldn't perform any validation of the result and shouldn't report any diagnostics for it if the type it is called for doesn't implement the interface, unless the mentioned operations affect resulting symbol.

### LINKED ISSUE #42340: master fails bootstrap build
A bootstrap build of the master branch by

> `.\Build.cmd -bootstrap -bootstrapConfiguration:Debug`

fails

```none
R:\roslyn\artifacts\Bootstrap\tasks\net472\Microsoft.CSharp.Core.targets(59,5): error MSB6006: "csc.exe" exited with code 1. [R:\roslyn\src\Compilers\Core\Portable\Microsoft.CodeAnalysis.csproj]
```

@jcouv diagnosed this to an assertion failure

```none
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.GetImplementsLocation(NamedTypeSymbol implementedInterface)
       at Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol.GetImplicitImplementationDiagnosticLocation(Symbol interfaceMember, TypeSymbol implementingType, Symbol member)
       at Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol.<>c.<CheckNullableReferenceTypeMismatchOnImplementingMember>b__95_2(DiagnosticBag diagnostics, MethodSymbol implementedMethod, MethodSymbol implementingMethod, ParameterSymbol implementingParameter, Boolean blameAttributes, ValueTuple`2 arg)
       at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.CheckValidNullableMethodOverride[TArg](CSharpCompilation compilation, MethodSymbol overriddenMethod, MethodSymbol overridingMethod, DiagnosticBag diagnostics, ReportMismatchinReturnType`1 reportMismatchInReturnType, ReportMismatchInParameterType`1 reportMismatchInParameterType, TArg extraArgument)
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #42340.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@gafter, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@gafter, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 1]
```diff
@@ -925,40 +934,47 @@ private SymbolAndDiagnostics ComputeImplementationAndDiagnosticsForInterfaceMemb
 
 #if !DEBUG
             // Don't optimize in DEBUG for better coverage for the GetInterfaceLocation function. 
-            if (useSiteDiagnostics != null)
+            if (useSiteDiagnostics != null && implementingTypeImplementsInterface)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: Does this make it so we add the `useSiteDiagnostics` in a release build even if `implementingTypeImplementsInterface` is `false`?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: > however it still seems like there might be a behavior inconsistency between debug and release builds.

That behavior difference is not observable, unless there is a bug in the method within the ```if``` statement. We intentionally stress them in DEBUG. There are other examples where we do extra work in DEBUG for the sake of extra validation.

---
In reply to: [393223483](https://github.com/dotnet/roslyn/pull/42405#discussion_r393223483) [](ancestors = 393223483)

**Pull request merged**